### PR TITLE
Student records across modules

### LINKED ISSUE #122: Student records of similar submissions across modules and assignments
It was suggested to have direct link access to the pair of submissions for "suspected plagiarism" and "confirmed plagiarism" past records.

### PR DESCRIPTION
**Pull request by nguyen-mai-huong**
Resolves #122 

This PR aims to help users with teaching duty to view the student's past "similar" submissions when he/she currently assesses a case of plagiarism. These past submissions could be accessed as below.

122-local-test

Related to: PR #83

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -80,13 +80,24 @@ def authenticate_actions_for_role(role, opts={})
       course = opts[:course]
   
       # Get current user's role
-      current_role = course.membership_for_user(@user).role
+      membership = course.membership_for_user(@user)
```
Comment (knmnyn) [inline app/controllers/application_controller.rb:83]: Changed all of the role variables in all files the same way?  Do you have documentation for why you changed the role?  (Just asking for future devs).

File: app/controllers/assignments_controller.rb [hunk 2]
```diff
@@ -43,13 +43,11 @@ class AssignmentsController < ApplicationController
...
   def index
-    
     # check whether is it a guest user 
-    @user = User.find_by_id(session[:user_id]) 
-    @guest_user = GuestUsersDetail.find_by_user_id(@user.id)
+    @guest_details = @course.find_guest_user_details(session[:user_id])
     # obtain assignment to be shown if is a guest user
-    if (@guest_user)
-      @assignment_to_be_shown = @guest_user.assignment_id
+    if @guest_details.length > 0
```
Comment (knmnyn) [inline app/controllers/assignments_controller.rb:49]: What is guest_details.length? Number of assignments that the guest user participates in?  Can you document this somewhere?

File: app/controllers/submission_logs_controller.rb [hunk 3]
```diff
@@ -17,7 +17,9 @@
 
 class SubmissionLogsController < ApplicationController
   before_action { |controller|
-    if params[:assignment_id]
+    if params[:assignment_id] and params[:submission_similarity_id]
+      controller.send :authenticate_custom_actions_for_teaching_staff, only: [:view_similarity]
```
Comment (knmnyn) [inline app/controllers/submission_logs_controller.rb:21]: Do you have specific documentation for what constitutes "teaching_staff" somewhere in your roles documentation?  (this seems like a very detailed controller function, so I want to know where all of the roles and documentation are).

File: app/models/guest_users_detail.rb [hunk 4]
```diff
@@ -1,2 +1,4 @@
 class GuestUsersDetail < ApplicationRecord
```
Comment (knmnyn) [inline app/models/guest_users_detail.rb:1]: Why is this called Details?  Should it be called something more like membership or record?

File: app/services/courses_service.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
+class CoursesService
+  def self.create_guest_membership(course, user)
+    membership = UserCourseMembership.new { |m|
+      m.role = 3
```
Comment (knmnyn) [inline app/services/courses_service.rb:4]: Please document all constants in an appropriate place and cross-reference it here.

File: app/views/assignments/index.html.erb [hunk 6]
```diff
@@ -17,7 +17,7 @@
 
 <% if @assignments.empty? %>
   <p class="none_created">No Assignment has been created.</p>
-<% elsif @guest_user %>
+<% elsif @guest_details.length > 0 %>
```
Comment (knmnyn) [inline app/views/assignments/index.html.erb:20]: Here, guest_details.length is later in your view described as "processed assignments", so does this mean the details are "processed assignments" (as opposed to just assignments)?

**Review by knmnyn — COMMENTED** [review]
Needs some work to standardize your terminology and document so that future devs can understand the logic.

**Review by knmnyn — APPROVED** [review]
LGTM.  Please review any comments that were made (I forgot whether I made any).

**Comment by knmnyn** [comment]
Please go ahead and merge when you're ready.

File: app/controllers/application_controller.rb [hunk 7]
```diff
@@ -80,13 +80,24 @@ def authenticate_actions_for_role(role, opts={})
...
       
       # Check if we need to authenticate
-      if current_role == role
-        unless opts[:only].include? action_name.intern
-          redirect_to( { controller: "announcements", action: "index" }, alert: "You do not have access to the url \"#{request.env['REQUEST_URI']}\". Please contact the administrator for more information.")
-        end
+      if membership.nil? or (membership and membership.role == role and !opts[:only].include? action_name.intern)
```
Comment (nguyen-mai-huong) [inline app/controllers/application_controller.rb:86]: The checking for role is resumed here.

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (nguyen-mai-huong) [reply app/controllers/application_controller.rb:83]: Nope, I just got membership at this line to avoid Nil Exception.

**Pull request merged**